actors/miner/src/lib.rs

```diff
@@ -1950,6 +1958,283 @@ impl Actor {
+                            sector.seal_rand_epoch,
+                            &entropy,
+                        )?
+                        .into(),
+                    ),
+                    interactive_randomness: Randomness(
+                        cid_to_replica_commitment_v1(&sector.sealed_cid)
+                            .map_err(|e| {
+                                actor_error!(
+                                    illegal_argument,
+                                    "invalid sealed CID {}: {}",
+                                    sector.sealed_cid,
+                                    e
+                                )
+                            })?
+                            .to_vec(),
+                    ),
+                    sealed_cid: sector.sealed_cid,
+                    unsealed_cid,
+                })
+            })
+            .collect::<Result<Vec<_>, _>>()?;
+
+        verify_aggregate_seal(
+            rt,
+            // All the proof inputs, even for invalid pre-commits,
+            // must be provided as witnesses to the aggregate proof.
+            &proof_inputs,
+            params.sectors[0].sealer_id,
+            params.seal_proof_type,
+            params.aggregate_proof_type,
+            &params.aggregate_proof,
+        )?;
+
+        let base_sector_power = base_power_for_sector(info.sector_size);
+
+        let rew = request_current_epoch_block_reward(rt)?;
+        let pwr = request_current_total_power(rt)?;
+        let circulating_supply = rt.total_fil_circ_supply();
+        let pledge_inputs = NetworkPledgeInputs {
+            network_qap: pwr.quality_adj_power_smoothed,
+            network_baseline: rew.this_epoch_baseline_power,
+            circulating_supply,
+            epoch_reward: rew.this_epoch_reward_smoothed,
+        };
+
+        let day_reward = expected_reward_for_power(
+            &pledge_inputs.epoch_reward,
+            &pledge_inputs.network_qap,
+            &base_sector_power,
+            fil_actors_runtime::EPOCHS_IN_DAY,
+        );
+
+        let storage_pledge = expected_reward_for_power(
+            &pledge_inputs.epoch_reward,
+            &pledge_inputs.network_qap,
+            &base_sector_power,
+            INITIAL_PLEDGE_PROJECTION_PERIOD,
+        );
+
+        let initial_pledge = initial_pledge_for_power(
+            &base_sector_power,
+            &pledge_inputs.network_baseline,
+            &pledge_inputs.epoch_reward,
+            &pledge_inputs.network_qap,
+            &pledge_inputs.circulating_supply,
+        );
+
+        let sectors_to_add = params
+            .sectors
+            .iter()
+            .map(|sector| SectorOnChainInfo {
+                sector_number: sector.sector_number,
+                seal_proof: params.seal_proof_type,
+                sealed_cid: sector.sealed_cid,
+                deprecated_deal_ids: vec![],
+                expiration: sector.expiration,
```

Comment:
`expiration` isn't checked anywhere in this method, we should have a call to `validate_expiration` for each sector earlier but instead of activation epoch in that call we should use current epoch (I think).